Closes #2724 Feature/batch duplicate search merge

MuchQuak opened the pull request:
# Issue #2724

# Summary
There is a lot going on in this pull request and so I have tried my best to get good working descriptions of what is being added and why. 

I had cherry picked most of these features into separate branches to pull requests to get general approval for the wider tools added, but due to security stuff going on I decided it was best to just forge ahead and get the detail review done here because it was blocking this tool's development and we currently don't have a process for proposing code improvements outside just doing them during feature development.

## Duplicate Copy Tool
Tool for copying georeference data between linked duplicates.

Currently copies the following fields (Currently Always Overrides):
-  `decimalLatitude`
-  `decimalLongitude`
- `geodeticDatum`
- `footprintWKT`
- `coordinateUncertaintyInMeters`
- `georeferencedBy`
- `georeferenceRemarks`
- `georeferenceSources`
- `georeferenceProtocol`
- `georeferenceVerificationStatus`

### Usage
Tool differentiates records into two categories `Targets` and `Options`. Targets are simply linked duplicates from the collection that matches the passed in `collId`. Options are linked duplicates of this targets and can be cyclical meaning if two targets are options of each other it will result in them showing up as targets and  options.

To use the tool search with the custom query to filter duplicate options based on field values. There are also two toggles which work like the following.

`Only show specimens from my collection without latitude and longitude`  toggle will limit the results to targets that are missing either latitude or longitude

`Only show duplicates with georeferences different than target specimen` toggle will limit the options to only show records that have at least one field that differs from the target record. This is so that collection managers can filter out options that have nothing to copy. This causes query is generally to be very slow but could save someone time if the search results are cluttered with not useful options.

There is also an option to filter by collections via the `Filter Collections` button which uses html `dialog` to render a popup with the collection checkbox form similarly seen in the collection search page. Selecting no options will result in  no filters being applied and thus searching all sections. This is also the case for selecting all collections and maybe this option is not actually needed?  Further details about the implementation are below in Collection Form.

## Custom Query Class
Added to decouple custom query inputs from `collections/editor/includes/queryform.php`. Needed to decouple because this copy tool needs to search outside of active collection. I think getting this integrated into `queryform.php` would to be helpful to ensure this ui component is reusable.

### Reasons
- Consolidates custom query request parsing and query construction logic
- Currently only support occurrence fields, but could be expanding as needed
- Moves HTML output to its own include in order to manage complexity with a wrapper to pass arguments to which allows for more control than only include.

### Usage

```php
// Defaults to use occurrence fields  because that is all it  is used for
CustomQuery::renderCustomInputs()

// But also allows for overrides in the case you  want to limit the fields.
CustomQuery::renderCustomInputs(['fieldName' => 'fieldTranslation' /*...*/])
```

##  Language Utility
Added Language Utility assist language loading

### Reasons
- Consolidate logic so all language loading can go through  the one function
- Ease of use,  newer or outside folks wouldn't be privy to the "right way" of loading these
- Because loading language files is easy breaking them apart is the better solutions than repeating the old pattern

### Usage
Single Loading
```php
// Loads language file at $SERVER_ROOT/content/lang/collections/sharedterms.[$LANG_TAG].php
Language::load('collections/sharedterms');
```

Multi loading
```php
Language::load(['collections/sharedterms', 'collections/editor/batchDuplicateGeorefCopy']);
```
## User Utility Class
Adds class to store user logic such as checking if they are a collection admin of given collection id.

Currently we have a lot of pages that have similar/same code for checking if user has permissions then using the `$isEditor` variable to alter the display or render that the user does not have permissions. 

This utilities class just consolidates this logic for checking if a user as authentication level (Currently only for collections because that is the only way it is used). It also provides an interface for checking the bool of an auth check with a fail rendering a separate access denied page and exit the process. 
### Reasons
- Currently copy pasting security logic instead of using functions and in light of recent security issues I think it makes sense to start consolidating this logic.
- Auth checks that result in rendering access denied still often do some database calls prior or logic processing or does some form of wrapping around `$isEditor` to prevent unnecessary calls. Adding utility to load separate access denied page prior to logic avoids all these issues and mitigates improper logic checks from downstream code of the auth check.

### Usage
Currently only collection admin check and accessDenied render check
```php
// Loads collId from request
$collId = array_key_exists('collid',$_REQUEST) && is_numeric($_REQUEST['collid'])? intval($_REQUEST['collid']):0;

// If user isn't authorized it will include "$SERVER_ROOT . '/includes/accessDenied.php'" and exit php process
UserUtil::isCollectionAdminOrDenyAcess($collId);
// Means that all code after is secure which cleans up logic checks here

```
Alternatively you can call them apart for flexibility
```php
UserUtil::authorizedOrDenyAccess(UserUtil::isCollectionAdmin($collId));
```

## Collections Form
New Include that is the collection checkboxes which mean this form is used in 3 different places 3 different ways.

### Reasons
- The `harvestparams.php` form could not be repurposed
- The new search collections selector include was extremely coupled to unspecified global variables and general structure could be simplified away from being 3 separate includes.

### Usage
It is just an include that requires not outside global variables. Downside is that a extra query may occur if all collections are needed outside of this element. Solution is to have some sort of session store for common query calls or allow pass through.

```php
include(__DIR__ . '/includes/collectionForm.php')
```

## Breadcrumbs
Something like this would help standardize existing pages that reuse component templates.

### Reasons
I don't' think we should go too crazy with it but it could seriously help manage simple dynamic components that are being reused in many places. This would help extra in the case of breadcrumbs because we can just import language tags used in breadcrumbs from the usage point meaning 'HOME' wouldn't be in a bunch of translations files.

If folks agree I can move the existing breadcrumbs over to the other pages.

### Usage
```php
breadcrumbs([
	$LANG['HOME'] => '../../index.php',
	$LANG['COL_MGMNT'] => '../misc/collprofiles.php?emode=1&collid=' . $collId,
	$LANG['BATCH_DUPLICATE_HARVESTER'],
])
```

## Duplicate Search
- Adjust  styles to fix display issue that I ran into.

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [x] Hotfixes should be branched off of the `master` branch and PR'd using the **merge** option (not squashed) into the `hotfix` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [x] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [x] Comment which GitHub issue(s), if any does this PR address
- [x] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [x] It is the code author's responsibility to merge their own pull request after it has been approved
- [x] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [x] If this PR represents a merge into the `hotfix` branch, remember to use the **merge** option (i.e., no squash).
- [x] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [x] If this PR represents a merge from the `hotfix` branch into the `master` branch use the **squash & merge** option
  - [x] a subsequent PR from `master` into `Development` should be made with the **merge** option (i.e., no squash).
  - [x] **Immediately** delete the `hotfix` branch and create a new `hotfix` branch
  - [x] increment the Symbiota version number in the symbase.php file and commit to the `hotfix` branch.
- [x] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [x] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

MuchQuak committed:
Proposal to have components folder (Breadcrumb Example)

# Summary
Something like this would help standardize existing pages that reuse
component templates.

I dont' think we should go too crazy with it but it could seriously help
manage simple dynamic components that are being reused in many places.

MuchQuak committed:
prototype commit

MuchQuak committed:
progess

MuchQuak committed:
Add ignores and get copy logic working

MuchQuak committed:
Create Helper Class and Seperate Include Custom Query Builder

# Summary
Seperates Custom Query Builder into its own include for easier reuse
uncoupled with query builder. Does not modify existing query form code
as I was planning on getting this the couple Duplicate Data Haravester
to development before.

This big enough change that I would like feedback and sanity checks
before moving forward.

MuchQuak committed:
Create utility class to standarize lang file loading

# Summary
Currently we use a kind of awkwardly sort of repeated pattern to load
lang files.

Something like this was easy to write and would save some folks time.
Especially if we change where the language files are stored or how they
are loaded.

MuchQuak committed:
Language Loader Util in CustomQuery Class

MuchQuak committed:
Change Term options to operator options and add operators

MuchQuak committed:
Add filters in getCustomValues so only valid values are returned

MuchQuak committed:
Add build customWhere helper to CustomQuery class

MuchQuak committed:
Fix query performance on bigger collections

MuchQuak committed:
Add distinct of fix duplicate occid issue

MuchQuak committed:
Add public occurrence link

MuchQuak committed:
Hook up custom query searching

MuchQuak committed:
Saving progress on getting modular duplicate search working

MuchQuak committed:
Fix custom query issue with IS_NULL and NOT_NULL operators

MuchQuak committed:
Get new duplicate searches rendering properly

MuchQuak committed:
Add order by and limit to duplicate cluster view

MuchQuak committed:
Add collid back to ignores

MuchQuak committed:
Rebuilding to have segmented search

MuchQuak committed:
Diff results so only applicable changes can be used

MuchQuak committed:
wrap check to ensure order of execution

MuchQuak committed:
fixup! Diff results so only applicable changes can be used

MuchQuak committed:
rename targetsOccids to targetOccids

MuchQuak committed:
Try oneHarvest

MuchQuak committed:
Add empty case getOccurrences

MuchQuak committed:
Fix $CLIENT_ROOT abs link

MuchQuak committed:
make record link create a new window

MuchQuak committed:
Fix less than and greate than

MuchQuak committed:
Adds missingLatLng search option and exact matches option

MuchQuak committed:
Add Include for collection form detached from other other stateful dependencies

MuchQuak committed:
prototype collection linkage

MuchQuak committed:
remove echo

MuchQuak committed:
clean up dead code

MuchQuak committed:
change wording missingLatLng

MuchQuak committed:
Sort Collections inlcude

MuchQuak committed:
Add inst and col code to end of collection name

MuchQuak committed:
Add more info links

MuchQuak committed:
Add back copyInfo call

MuchQuak committed:
Add subheader

MuchQuak committed:
Move subheader margin closer to relevant content

MuchQuak committed:
Reword hideExactMatches

MuchQuak committed:
Add collection saving betweenr requests

MuchQuak committed:
Add array key check for collection checking

MuchQuak committed:
Add basic pagination

MuchQuak committed:
Fix client root

MuchQuak committed:
row render link fix

MuchQuak committed:
rename batchDuplicateMerge to batchDuplicateGeorefCopy

Renames file to be more accurate to what it is actually doing

MuchQuak committed:
add link and lang en tag to duplicate manager

MuchQuak committed:
Remove used code left from prototyping

MuchQuak committed:
Add User util and access denied page for handling admin errors

MuchQuak committed:
Add shared term lang tags

MuchQuak committed:
Change name of checkAccess to be more clear it will deny access

MuchQuak committed:
LoadLanguage After auth check

MuchQuak committed:
remove removed lang tag

MuchQuak committed:
Add back breadcrumbs proposal

MuchQuak committed:
create language file for georefCopy tool

MuchQuak committed:
Add shared term for close

MuchQuak committed:
Add language translations

MuchQuak committed:
add lang tag for title and search

MuchQuak committed:
Add lang tag for search criteria subheader

MuchQuak committed:
Add record permissions

themerekat reviewed (changes requested):
Working great so far! One critical missing component: someone who does not have rare species reader or editor/admin access to a record that has locality security (omoccurrences.recordsecurity = 1) can see the lat/long values for that record. For example, I'm logged in as someone who ONLY has access to the SBBG collection. This record has locality security:
<img width="935" height="860" alt="image" src="https://github.com/user-attachments/assets/c9037e36-4c03-434e-b45a-a63f90d5c7a3" />

But that security is not applied in this tool:
<img width="1821" height="132" alt="image" src="https://github.com/user-attachments/assets/1796e2af-e272-427d-ac46-b41442199386" />

I think the most straightforward way to deal with this is to exclude any records that have localitysecurity = 1, regardless of whether the user can view them, because copying those redacted records into your own collection would effectively make that data un-redacted.

themerekat commented:
Another thing we should ask the users about: do we want to append something automatically to the georeferenceRemarks when we copy the data? For example, "copied from duplicate [institutioncode]: [catalognumber]" ? I did this when I created the standalone tool, and it was nice to reference.

ljwalker commented:
FWIW I had a similar thought and think this would be really helpful as a reference!
> do we want to append something automatically to the georeferenceRemarks when we copy the data?

Perhaps with a date stamp: `copied from duplicate [institutioncode]: [catalognumber] on [date]`

themerekat commented:
> FWIW I had a similar thought and think this would be really helpful as a reference!
> 
> > do we want to append something automatically to the georeferenceRemarks when we copy the data?
> 
> Perhaps with a date stamp: `copied from duplicate [institutioncode]: [catalognumber] on [date]`

We really should implement the dateGeoreferenced field, tbh, but this is a good option for now!

MuchQuak committed:
Hide records with locality security

MuchQuak commented:
@themerekat Code is on dev002 for testing and security flags have been updated to include locality security. Just a heads up locality security is now being saved into `recordSecurity` under the value of 1.

themerekat commented:
## Feedback from users (after meeting)

1. Add the fields "collector" (recordedBy), "collector number" (recordnumber), and date (eventDate) to the printout of available fields. They don't need to take up tons of space, but the users will likely want to examine at least the collector numbers and dates to ensure they aren't "s.n" or "0000-00-00".

2. Add a checkbox/button at the top of the printout that automatically checks all of the boxes in the currently visible printout. If possible, it would be nice for it to NOT check boxes for clusters that have more than one matching duplicate. Here's an example printout of what that would look like for these records:
<img width="520" height="558" alt="image" src="https://github.com/user-attachments/assets/d2495b09-3ba4-4d8b-9143-f133b4a27ce1" />

3. When duplicate data is copied, prepend the field "georeferenceRemarks" with the following text: "copied from duplicate [institutionCode] [catalognumber OR othercatalognumbers, only if catalognumber is null]"

jbest commented:
If I understand correctly, the batch georef duplicate tool can only match specimens that have previously been clustered using the "[Batch link specimen duplicates](https://dev002.symbiota.org/portal/collections/datasets/duplicatemanager.php?collid=34&action=batchlinkdupes) tool. If that's the case, it would be helpful to include somewhere in the Batch Geo Dups page the last date of specimen clustering so if the date wasn't recent, the user could choose to run another batch cluster. Maybe something like "Last specimen cluster performed 2025-09-12. Any records updated or added since that time will not be included in this duplicate discovery."

jbest commented:
When I first navigate to the Batch Georef Dup tool, there are already results listed even though I haven't performed a filter or search (in this session). It's not clear what is listed, but I assume all potential dups? Perhaps there should be no results shown until the user selects search (with optional filters applied).

jbest commented:
The checkboxes in each duplicate georef cluster are single choice only so should be changed to radio buttons. But as I write this I realize you need to be able to uncheck if you decide a georef copy shouldn't be done. Checkboxes are probably OK for this case.

jbest commented:
It would be ideal to have a filter option to return only the duplicate georef results that have one and only one georef source. That would allow the user to do a first pass that matches the only choices, then the user can go back with a second pass that returns dup matches that have more than one source. The user can then pick and choose which ones they want. (You may have this request in meeting notes but I didn't see it here. Apologies if this is a duplicate request.)

jbest commented:
I think the duplicate geo tool checkbox for "Only show specimens from my collection without latitude and longitude" should be checked by default because, at least in my primary intended use of the workflow, the goal is to populate records that entirely lack georefs, not to update or improve existing records. Alternatively, the default search behavior could be changed to only include records without georef, and change the filter option to something like "Show specimens from my collection that have existing longitude and latitude."

jbest commented:
After a geo ref duplicate result is copied by the user, the list of updated records is displayed followed by the updated search results. Currently clicking on any of the links to the updated record opens the page in the same tab, losing the context of the copy results and georef dup tool. Ideally the updated record links would open in a new tab or window.

jbest commented:
After a geo ref duplicate result is copied by the user, the list of updated records is displayed followed by the updated search results but the previous filters used are not retained. This requires the user to enter them again and perform a new search.

MuchQuak committed:
add collector and collector number to display

MuchQuak committed:
enable auto checking by default

themerekat commented:
Request from Lauren H: add locality (abbreviated version, probably) to the duplicate display

At this point, I wonder if there needs to be a way to toggle which fields you see and don't see. Not sure if that would be a huge pain to implement.

MuchQuak committed:
Enforce autocheck behavior to only function on differing collections

This removes edge case issues where records pointing at each other will
get merge into each other causing 1 to have the merge and the other one
to not which is odd behavior.

MuchQuak committed:
Redo autochecking to check to enable instead of disable

MuchQuak committed:
add locality to table

MuchQuak committed:
fix pagination text to reflect proper action

MuchQuak committed:
Create sticky session variables that work the same after copying data

MuchQuak committed:
Updated links always open in new tab

MuchQuak committed:
Add note to georeference remarks that shows where record has pull in duplicate from

MuchQuak committed:
Change duplicates missintg lat lng values to the default and cleanup what values are default with a form message passing

MuchQuak committed:
Disable search by default. Users must click search to see results

MuchQuak committed:
basic non stick field toggling

MuchQuak committed:
upgrade field toggling to be sticky to requests and rework alias to allow any field to be aliased as any text

MuchQuak commented:
@themerekat I have made updates for all feedback minus being able to see the last performed duplicate linking. Implementing this as is is going to incur a 5+ second load tax on many collections using this tool. I can still add it, but I would suggest splitting it into separate enhancement issue as the work to get these changes in the batch linking tool itself.

themerekat commented:
**DONE**

Suggested change to this verbiage:
<img width="880" height="32" alt="image" src="https://github.com/user-attachments/assets/6def4636-5a4e-4875-b18e-4eec1bfd3e36" />

To "Auto-select duplicates with only one option (note: duplicates within your collection will not be auto-selected)"

themerekat committed:
Change field order, add advisory to batch link dupes

themerekat committed:
Add missing translations

themerekat committed:
Uncheck auto-select duplicates option

themerekat reviewed (approved):
The tool works great! I made some minor UI changes and added missing translations.

MuchQuak commented:
> The tool works great! I made some minor UI changes and added missing translations.

Thanks, I going to wait until the rest of the team gets some review in then I will merge it.

Atticus29 commented on classes/CustomQuery.php:
```diff
@@ -0,0 +1,244 @@
+				'predicate' => fn($v) => array_key_exists($v, self::OPERATOR_OPTIONS)
+			],
+			'q_customvalue' => [
+				'field' => 'value',
```
Not sure how customfield is different from customvalue and why customvalue is missing a predicate.

MuchQuak replied:
Predicate is just a true or false filter. Could change name to filter if clear. Custom field is a restricted field where only certain values are excepted. Custom value has no such restrictions. If both aren't populated then it gets thrown out anyway.

Atticus29 commented on classes/CustomQuery.php:
```diff
@@ -0,0 +1,244 @@
+			],
+			'q_customopenparen' => [
+				'field' => 'openparen',
+				'predicate' => fn($v) => preg_match('/^\({1,3}$/', $v)
```
Why only 1-3?

MuchQuak replied:
This was the limitations enacted by the editor version so keeping it consistent so that it can be piped there without error.

Atticus29 commented on classes/CustomQuery.php:
```diff
@@ -0,0 +1,244 @@
+			],
+			'q_customcloseparen' => [
+				'field' => 'closeparen',
+				'predicate' => fn($v) => preg_match('/^\){1,3}$/', $v)
```
Why only 1-3?

MuchQuak replied:
Same as open param

Atticus29 commented on classes/utilities/Language.php:
```diff
@@ -0,0 +1,30 @@
+	 * @param string|array $path Filepath to load language files for. Exclude .php extension
+	 * @return void
+	 **/
+	static function load(mixed $path): void {
```
maybe better to use `array|string` instead of `mixed`?

MuchQuak replied:
We cannot use string|array because of backwards compat

MuchQuak replied:
see https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.composite

Atticus29 commented on classes/CustomQuery.php:
```diff
@@ -0,0 +1,244 @@
+			'verbatimEventDate'=>$LANG['VERBATIM_DATE'],
+			'verbatimDepth'=>$LANG['VERBATIM_DEPTH'],
+			'verbatimElevation'=>$LANG['VERBATIM_ELE'],
+			'waterbody'=> $LANG['WATER_BODY']
```
We'll somehow have to remember to update this if/when we add new columns to omoccurrences. Maybe we can add something like updating this to the PR checklist if new columns are added?

MuchQuak replied:
Couple things. If this replaces how we did before it isn't making anything worse, but putting the original logic in a package that is easier to reuse. If all the custom input logic was handled here, I think it would be pretty reasonable to remember to add a one liner to an array after having to also remember to write all the search logic, and add other search inputs to the other search pages.

I don't disagree that there isn't some code duplication here that will cause some maintenance overhead. I would be more in favor of making an github issue with a tech debt label that is up for grabs instead of another checklist item.

Atticus29 commented on classes/CustomQuery.php:
```diff
@@ -0,0 +1,244 @@
+			$CUSTOM_FIELDS = $customFields;
+		}
+
+		$MAX_CUSTOM_INPUTS = self::MAX_CUSTOM_INPUTS;
```
Why is there a limit to 8?

MuchQuak replied:
The limit has always been 8. This is because the inputs hidden so there must be a fixed limit. 8 is what was chosen before so it has been kept that way.

Atticus29 commented on classes/utilities/UserUtil.php:
```diff
@@ -0,0 +1,44 @@
+
+	/**
+	 * Checks if user is a collectionAdmin for given collId
+	 * If then are not then php will include accessDenied.php
```
*they

MuchQuak replied:
Thanks for the catch

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+include_once($SERVER_ROOT . '/classes/Sanitize.php');
+include_once($SERVER_ROOT . '/classes/CustomQuery.php');
+
+// Other none copied fields selected out of occurrence table
```
*non?

MuchQuak replied:
Thanks for the catch

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+include_once($SERVER_ROOT . '/classes/CustomQuery.php');
+
+// Other none copied fields selected out of occurrence table
+$fields = [
```
maybe just call this $nonCopiedFields or something to self-document?

MuchQuak replied:
Adjusted name and comment

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+<?php
```
This file almost seems like it should be divided into a class file and a view. Lots of method definitions and config that might could be divided out and help with line count.

MuchQuak replied:
(Quick note first) I would be willing to make an issue to address this separately. There is already a bunch of things in this pull request that have blown past the scope of the feature being added.

I generally agree with the sentiment that this logic mostly should live elsewhere, but moving it to a separate file is not solving that issue it is just pushing it under the bed.

There is a lot of underling issues, one of which is we don't have a model system or any kind of generalized data fetching for common things like collections and occurrences. Often times integraiting with existing solutions reliant on 3 deep class inheritance is a nightmare when you just want to build a query. We don't need more classes we need better ones.

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+	'collections/list'
+]);
+
+$start = array_key_exists('start',$_REQUEST)?$_REQUEST['start']:0;
```
Add spacing for ternary operators.

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+	OccurrenceUtil::appendFullProtectionSQL();
+
+	// Don't process records with hidden locality information
+	$sql .= ' AND o.recordSecurity != 1 ';
```
Do we also need to handle o.recordSecurity != 5?

MuchQuak replied:
This is already added via appendFullProtectionSQL

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+	$options[$duplicateId][$occid] = $option;
+}
+
+function getUniqueOptionCount($options, $targetOccid) {
```
I imagine that PHP has more efficient ways of counting unique, but this doesn't feel vital.

MuchQuak replied:
I would be all ears if you can think of a better way of doing this. The most this gets run is 100 times per load which is pretty small overhead so not a huge deal.

Atticus29 commented on collections/editor/batchDuplicateGeorefCopy.php:
```diff
@@ -0,0 +1,543 @@
+
+				<div style="display:flex; gap:1rem;margin-bottom:1rem;">
+					<dialog id="table_toggle_dialog" style="min-width: 900px;">
+						<div style="display:flex; align-items: center; gap: 0.5rem; margin-bottom: 1rem;">
```
There's a mix of inline styling and in-file styling. Not a big deal, but worth mentioning.

MuchQuak replied:
What is the alternative here? I could make a selector that is used once in this file, but I don't really understand what benefit that generates. I separated the more complex styles into selectors, kept the formatting and spacing to be inline (there did not seem to be an obvious class to use for this), and reused classes where I could.

Atticus29 commented on collections/editor/includes/collectionForm.php:
```diff
@@ -0,0 +1,187 @@
+<?php
```
I'm interested to see how this would be implemented in the other pages that require a collection form.

>  ... Solution is to have some sort of session store for common query calls or allow pass through.

I think you showed us during a dev meeting how this might work, but I don't remember it well.

Atticus29 replied:
In other words, I'm comfortable with this replacing the previous version, and I think it's awesome how it's only 187 lines. I just worry that repurposing the search pages might prove non-trivial.

Atticus29 replied:
@themerekat  I think that it will also be important to test this on a portal that has collection categories as well as one that doesn't.

Atticus29 commented on collections/editor/includes/customInput.php:
```diff
@@ -0,0 +1,119 @@
+<?php
+global $LANG, $LANG_TAG, $SERVER_ROOT;
+
+if($LANG_TAG != 'en' && file_exists($SERVER_ROOT.'/content/lang/collections/editor/includes/queryform.'.$LANG_TAG.'.php')) include_once($SERVER_ROOT.'/content/lang/collections/editor/includes/queryform.'.$LANG_TAG.'.php');
```
use Language::load() here?

MuchQuak replied:
Makes sense, I think is used to but I decoupled them for separate pr then merged them back in. I will get them using the loader.

Atticus29 reviewed (commented):
Most of my comments are cosmetic, but there might be a thing or two in there that need changing if warranted.

MuchQuak committed:
fix doc typo

MuchQuak committed:
other doc typo

MuchQuak committed:
adjust naming of fields to be more specific

MuchQuak committed:
remove snake case

MuchQuak committed:
formating fix

MuchQuak committed:
Swap customInput to use languageLoader

MuchQuak committed:
Move Authgate to top

MuchQuak committed:
Reorganize file to be consts, request handling, function lib, search logic

MuchQuak committed:
Move breadcrumbs to a class under classes
